lib/Conversion/TorchToMhlo/ReductionOp.cpp

```diff
@@ -0,0 +1,566 @@
+  if (!initValue) return failure();
+  auto mhloReduceOp = rewriter.create<mhlo::ReduceOp>(
+      op.getLoc(), input, initValue, rewriter.getI64TensorAttr(dims));
+
+  Block &block = mhloReduceOp.body().emplaceBlock();
+  auto blockArgumentTy = RankedTensorType::get({}, inputTy.getElementType());
+
+  block.addArgument(blockArgumentTy, op->getLoc());
+  block.addArgument(blockArgumentTy, op->getLoc());
+
+  auto *firstArgument = block.args_begin();
+  auto secondArgument = block.args_rbegin();
+
+  {
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToStart(&block);
+    Value maxResult = rewriter.create<mhlo::MaxOp>(
+        op->getLoc(), blockArgumentTy, *firstArgument, *secondArgument);
+    rewriter.create<mhlo::ReturnOp>(op->getLoc(), maxResult);
+  }
+
+  rewriter.replaceOp(op, mhloReduceOp.getResults());
+  return success();
+}
+} // namespace
+
+// AtenSumDimIntListOp
+namespace {
+template <>
+LogicalResult ConvertAtenReductionOp<AtenSumDimIntListOp>::matchAndRewrite(
+    AtenSumDimIntListOp op, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+  Value input = adaptor.self();
+  auto inputTy = input.getType().dyn_cast<RankedTensorType>();
+  if (!inputTy) {
+    return rewriter.notifyMatchFailure(op, "only Tensor types supported in MHLO");
+  }
+  auto dtype = adaptor.dtype();
+  if (!dtype.getType().isa<Torch::NoneType>()) {
+    auto dstElemTy = getTypeConverter()
+                         ->convertType(op.getType())
+                         .template dyn_cast<RankedTensorType>()
+                         .getElementType();
+    input = rewriter.create<mhlo::ConvertOp>(op->getLoc(), input, dstElemTy);
+    inputTy = input.getType().dyn_cast<RankedTensorType>();
+  }
+  auto inputElemTy = inputTy.getElementType();
+  if (!inputElemTy.isIntOrFloat()) {
+    return op.emitError(
+        "Only floating-point or integer datatype legalization supported");
+  }
+
+  // Currently, (u)int8 dtype is not supported
+  if (inputElemTy.isa<mlir::IntegerType>() &&
+      inputElemTy.getIntOrFloatBitWidth() == 8) {
+    return rewriter.notifyMatchFailure(
+        op, "IntegerType with bitwidth 8 unsupported in convertion from "
+            "AtenSumDimIntListOp to MHLO");
+  }
+
+  SmallVector<int64_t> inputDims;
+  SmallVector<int64_t> dims;
+  if (!matchPattern(op.dim(), m_TorchConstantIntList(inputDims))) {
+    return rewriter.notifyMatchFailure(op, "non-int dim list unsupported");
+  }
+
+  for (auto d : inputDims) {
+    d = toPositiveDim(d, inputTy.getRank());
+    // Drop invalid dims
+    if (isValidDim(d, inputTy.getRank())) {
+      dims.push_back(d);
+    }
+  }
+
+  bool keepDim = false;
+  if (!matchPattern(op.keepdim(), m_TorchConstantBool(&keepDim))) {
+    return rewriter.notifyMatchFailure(op, "non-bool keepdim unsupported");
+  }
+  Value initValue =
+      createInitialValueForReduceOp(op, inputTy.getElementType(), rewriter);
+  if (!initValue) return failure();
```

Comment:
return the failure with an error message here?

Reply:
Function `createInitialValueForReduceOp` is responsible for throwing the error message.